File: alc/cgen.c
// QBE code generator for the alic compiler.
// (c) 2025 Warren Toomey, GPL3

#include <stdio.h>
#include <stdlib.h>
#include "alic.h"
#include "proto.h"

static int nexttemp = 1;	// Incrementing temporary number

// Allocate a QBE temporary
static int cgalloctemp(void) {
  return (++nexttemp);
}

// Generate a label
void cglabel(int l) {
  fprintf(Outfh, "@L%d\n", l);
}

// Generate a string literal
void cgstrlit(int label, char *val) {
  int i;
  fprintf(Outfh, "data $L%d = { b \"", label);

  // Interpret some control characters
  for (i=0; val[i] != 0; i++) {
    switch(val[i]) {
    case '\a': fprintf(Outfh, "\\a"); break;
    case '\b': fprintf(Outfh, "\\b"); break;
    case '\f': fprintf(Outfh, "\\f"); break;
    case '\n': fprintf(Outfh, "\\n"); break;
    case '\r': fprintf(Outfh, "\\r"); break;
    case '\t': fprintf(Outfh, "\\t"); break;
    case '\v': fprintf(Outfh, "\\v"); break;
    default:   fprintf(Outfh, "%c", val[i]);
    }
  }

  fprintf(Outfh, "\", b 0 }\n");

}

// Generate a jump to a label
void cgjump(int l) {
  fprintf(Outfh, "  jmp @L%d\n", l);
}

// Table of QBE type names used
// after the '=' sign in instructions
static char *qbe_typename[TY_FLT64 + 1] = {
  "", "w", "w", "w", "w", "l", "s", "d"
};

// Table of QBE type names used
// in store instructions
static char *qbe_storetypename[TY_FLT64 + 1] = {
  "", "b", "b", "h", "w", "l", "s", "d"
};

// Table of QBE type names used when loading.
// Second half represents unsigned types
static char *qbe_loadtypename[2 * (TY_FLT64 + 1)] = {
  "", "sb", "sb", "sh", "sw", "l", "s", "d",
  "", "ub", "ub", "uh", "uw", "l", "s", "d",
};

// Table of QBE type names used when extending.
// Second half represents unsigned types
static char *qbe_exttypename[2 * (TY_FLT64 + 1)] = {
  "", "sw", "sw", "sw", "sw", "sl", "s", "d",
  "", "uw", "uw", "uw", "uw", "ul", "s", "d",
};

// Return the QBE type that
// matches the given built-in type
static char *qbetype(Type * type) {
  int kind = type->kind;

  if (type->kind > TY_FLT64) fatal("not a built-in type");
  if (type->kind == TY_VOID) fatal("no QBE void type");
  return (qbe_typename[kind]);
}

// Ditto for stores
static char *qbe_storetype(Type * type) {
  int kind = type->kind;

  if (type->kind > TY_FLT64) fatal("not a built-in type");
  if (type->kind == TY_VOID) fatal("no QBE void type");
  return (qbe_storetypename[kind]);
}

// Ditto for loads, with signed knowledge
static char *qbe_loadtype(Type * type) {
  int kind = type->kind;

  if (type->kind > TY_FLT64) fatal("not a built-in type");
  if (type->kind == TY_VOID) fatal("no QBE void type");
  if (type->is_unsigned) kind += TY_FLT64 + 1;
  return (qbe_loadtypename[kind]);
}

// Ditto for extends, with signed knowledge
static char *qbe_exttype(Type * type) {
  int kind = type->kind;

  if (type->kind > TY_FLT64) fatal("not a built-in type");
  if (type->kind == TY_VOID) fatal("no QBE void type");
  if (type->is_unsigned) kind += TY_FLT64 + 1;
  return (qbe_exttypename[kind]);
}

// Print out the file preamble
void cg_file_preamble(void) {
}

// Print out the function preamble
void cg_func_preamble(Sym *func) {
  Sym *this;
  char *qtype;

  fprintf(Outfh, "export function $%s(", func->name);

  // Output the list of parameters
  for (this= func->memb; this != NULL; this= this->next) {
    // Get the parameter's type
    qtype= qbetype(this->type);
    fprintf(Outfh, "%s %%%s", qtype, this->name);

    // Print out any comma separator
    if (this->next != NULL)
      fprintf(Outfh, ", ");
  }

  fprintf(Outfh, ") {\n");
  fprintf(Outfh, "@START\n");
}

// Print out the function postamble
void cg_func_postamble(void) {
  fprintf(Outfh, "@END\n");
  fprintf(Outfh, "  ret\n");
  fprintf(Outfh, "}\n");
}

// Define a global symbol
void cgglobsym(Sym * s) {
  if (s == NULL) return;

  // Get the matching QBE type
  char *qtype = qbe_storetype(s->type);

  switch (s->type->kind) {
  case TY_FLT32:
  case TY_FLT64:
    fprintf(Outfh, "export data $%s = { %s %s_%f, }\n",
	    s->name, qtype, qtype, s->initval.dblval);
    break;
  default:
    fprintf(Outfh, "export data $%s = { %s %ld, }\n",
	    s->name, qtype, s->initval.intval);
  }
}

// Print out temporary's value using printf()
void cgprint(int label, int temp, Type *type) {
  // Get the matching QBE type
  char *qtype = qbetype(type);

  fprintf(Outfh, "  call $printf(l $L%d, %s %%.t%d)\n",
		label, qtype, temp);
}

// Load an integer literal value into a temporary.
// Return the number of the temporary.
int cgloadlit(Litval value, Type * type) {
  // Get a new temporary
  int t = cgalloctemp();

  // Get the matching QBE type
  char *qtype = qbetype(type);

  switch (type->kind) {
  case TY_FLT32:
  case TY_FLT64:
    fprintf(Outfh, "  %%.t%d =%s copy %s_%f\n", t, qtype, qtype,
	    value.dblval);
    break;
  default:
    fprintf(Outfh, "  %%.t%d =%s copy %ld\n", t, qtype, value.intval);
  }
  return (t);
}

// Perform a binary operation on two temporaries and
// return the number of the temporary with the result
static int cgbinop(int t1, int t2, char *op, Type * type) {
  // Get the matching QBE type
  char *qtype = qbetype(type);

  fprintf(Outfh, "  %%.t%d =%s %s %%.t%d, %%.t%d\n", t1, qtype, op, t1, t2);
  return (t1);
}

// Add two temporaries together and return
// the number of the temporary with the result
int cgadd(int t1, int t2, Type * type) {
  return (cgbinop(t1, t2, "add", type));
}

// Subtract the second temporary from the first and
// return the number of the temporary with the result
int cgsub(int t1, int t2, Type * type) {
  return (cgbinop(t1, t2, "sub", type));
}

// Multiply two temporaries together and return
// the number of the temporary with the result
int cgmul(int t1, int t2, Type * type) {
  return (cgbinop(t1, t2, "mul", type));
}

// Divide the first temporary by the second and
// return the number of the temporary with the result
int cgdiv(int t1, int t2, Type * type) {
  return (cgbinop(t1, t2, "div", type));
}

// Negate a temporary's value
int cgnegate(int t, Type * type) {
  fprintf(Outfh, "  %%.t%d =%s sub 0, %%.t%d\n", t, qbetype(type), t);
  return (t);
}

// List of QBE comparison operations. Add 6 for unsigned
static char *qbecmp[] = {
  "eq", "ne", "slt", "sgt", "sle", "sge",
  "eq", "ne", "slt", "sgt", "sle", "sge"
};

// Compare two temporaries and return the boolean result
int cgcompare(int op, int t1, int t2, Type * type) {
  // Get the matching QBE type
  char *qtype = qbetype(type);

  // Get the QBE comparison
  int offset = type->is_unsigned ? 6 : 0;
  char *cmpstr = qbecmp[op - A_EQ + offset];

  // Get a new temporary
  int t = cgalloctemp();

  fprintf(Outfh, "  %%.t%d =w c%s%s %%.t%d, %%.t%d\n",
	  t, cmpstr, qtype, t1, t2);
  return (t);
}

// Jump to the label if the value in t1 is zero
void cgjump_if_false(int t1, int label) {
  // Get a label for the next instruction
  int label2 = genlabel();

  fprintf(Outfh, "  jnz %%.t%d, @L%d, @L%d\n", t1, label2, label);
  cglabel(label2);
}

// Logically NOT a temporary's value
int cgnot(int t, Type * type) {
  // Get the matching QBE type
  char *qtype = qbetype(type);

  fprintf(Outfh, "  %%.t%d =%s ceq%s %%.t%d, 0\n", t, qtype, qtype, t);
  return (t);
}

// Invert a temporary's value
int cginvert(int t, Type * type) {
  fprintf(Outfh, "  %%.t%d =%s xor %%.t%d, -1\n", t, qbetype(type), t);
  return (t);
}

// Bitwise AND two temporaries together and return
// the number of the temporary with the result
int cgand(int t1, int t2, Type * type) {
  return (cgbinop(t1, t2, "and", type));
}

// Bitwise OR two temporaries together and return
// the number of the temporary with the result
int cgor(int t1, int t2, Type * type) {
  return (cgbinop(t1, t2, "or", type));
}

// Bitwise XOR two temporaries together and return
// the number of the temporary with the result
int cgxor(int t1, int t2, Type * type) {
  return (cgbinop(t1, t2, "xor", type));
}

// Shift left t1 by t2 bits
int cgshl(int t1, int t2, Type * type) {
  return (cgbinop(t1, t2, "shl", type));
}

// Shift right t1 by t2 bits
int cgshr(int t1, int t2, Type * type) {
  return (cgbinop(t1, t2, "shr", type));
}

// Load a value from a variable into a temporary.
// Return the number of the temporary.
int cgloadvar(Sym *sym) {
  // Allocate a new temporary
  int t = cgalloctemp();

  // Get the matching QBE type
  char *qloadtype = qbe_loadtype(sym->type);
  char *qtype = qbetype(sym->type);

  if (sym->has_addr)
    fprintf(Outfh, "  %%.t%d =%s load%s %%%s\n",
			t, qtype, qloadtype, sym->name);
  else
    fprintf(Outfh, "  %%.t%d =%s copy %%%s\n",
			t, qtype, sym->name);
  return (t);
}

void cgstorvar(int t, Type * exprtype, Sym * sym) {
  // Get the matching QBE type
  char *qtype = qbe_storetype(sym->type);

  if (sym->has_addr)
    fprintf(Outfh, "  store%s %%.t%d, %%%s\n", qtype, t, sym->name);
  else
    fprintf(Outfh, "  %%%s =%s copy %%.t%d\n",
			sym->name, qtype, t);
}

// Cast a temporary to have a new type
int cgcast(int t1, Type * type, Type * newtype) {
  // Allocate a new temporary
  int t2 = cgalloctemp();

  // As t1 is already word-sized,
  // we can upgrade the alic type for t1
  switch (type->kind) {
  case TY_BOOL:
  case TY_INT8:
  case TY_INT16:
    if (type->is_unsigned) type = ty_uint32;
    else		   type = ty_int32;
    break;
  default:
  }

  // Get the matching QBE types
  char *oldqtype = qbe_exttype(type);
  char *newqtype = qbetype(newtype);

  // Conversion from int to flt
  if (is_integer(type) && is_flonum(newtype)) {
    fprintf(Outfh, "  %%.t%d =%s %stof %%.t%d\n", t2, newqtype, oldqtype, t1);
    return (t2);
  }

  // Widening
  if (newtype->size > type->size) {
    switch (type->kind) {
    case TY_INT32:
      fprintf(Outfh, "  %%.t%d =%s ext%s %%.t%d\n",
	      t2, newqtype, oldqtype, t1);
      break;
    case TY_FLT32:
      fprintf(Outfh, "  %%.t%d =%s ext%s %%.t%d\n",
	      t2, newqtype, oldqtype, t1);
      break;
    default:
      fatal("Not sure how to widen from %s to %s\n",
	    get_typename(type), get_typename(newtype));
    }
    return (t2);
  }

  // Narrowing
  if (newtype->size < type->size) {
    switch (type->kind) {
    case TY_INT32:
      return (t1);
    default:
      fatal("Not sure how to narrow from %s to %s\n",
	    get_typename(type), get_typename(newtype));
    }
    return (t2);
  }

  // We didn't narrow or widen!
  return (t1);
}

// Add space for a local variable
void cgaddlocal(Type *type, Sym *sym) {
  // Get the type's size.
  // Make it at least 4 bytes in size
  // as QBE requires this for a variable
  // on the stack
  int size= (type->size < 4) ? 4 : type->size;

  fprintf(Outfh, "  %%%s =l alloc%d 1\n", sym->name, size);
}

// Call a function with the given symbol id.
// Return the temporary with the result
int cgcall(Sym *sym, int numargs, int *arglist, Type **typelist) {
  int rettemp;
  int i;

  // Get a new temporary for the return result
  rettemp = cgalloctemp();

  // Call the function
  if (sym->type == ty_void)
    fprintf(Outfh, "  call $%s(", sym->name);
  else
    fprintf(Outfh, "  %%.t%d =%s call $%s(",
	rettemp, qbetype(sym->type), sym->name);

  // Output the list of arguments
  for (i = 0; i < numargs; i++) {
    fprintf(Outfh, "%s %%.t%d", qbetype(typelist[i]), arglist[i]);
    if (i < numargs-1) fprintf(Outfh, ", ");
  }
  fprintf(Outfh, ")\n");

  return (rettemp);
}

File: alpy/cgen.py
from io import StringIO
from typing import List

from defs import DataType, TypeKind, Sym, Litval, ASTNodeType, fatal, quote_string
from strlits import strlit_processor


class CodeGenerator:
    qbe_type_names = [
        "", "w", "w", "w", "w", "l", "s", "d"
    ]
    qbe_store_type_names = [
        "", "b", "b", "h", "w", "l", "s", "d"
    ]
    qbe_load_type_names = [
        "", "sb", "sb", "sh", "sw", "l", "s", "d",
        "", "ub", "ub", "uh", "uw", "l", "s", "d"
    ]
    qbe_ext_type_names = [
        "", "sw", "sw", "sw", "sw", "sl", "s", "d",
        "", "uw", "uw", "uw", "uw", "ul", "s", "d"
    ]
    arithmetic_ops = {
        ASTNodeType.A_ADD: "add", ASTNodeType.A_SUBTRACT: "sub",
        ASTNodeType.A_MULTIPLY: "mul", ASTNodeType.A_DIVIDE: "div",
    }
    logical_ops = {
        ASTNodeType.A_AND: "and", ASTNodeType.A_OR: "or", ASTNodeType.A_XOR: "xor",
        ASTNodeType.A_LSHIFT: "shl", ASTNodeType.A_RSHIFT: "shr",
    }
    comparison_ops = {
        ASTNodeType.A_EQ: "eq", ASTNodeType.A_NE: "ne",
        ASTNodeType.A_LE: "le", ASTNodeType.A_LT: "lt",
        ASTNodeType.A_GE: "ge", ASTNodeType.A_GT: "gt",
    }

    def __init__(self):
        self.output = StringIO()
        self.next_temp = 1
        self.label_id = 1

    @staticmethod
    def check_type(val_type: DataType):
        if val_type.kind.value > TypeKind.TY_FLT64.value:
            fatal("not a built-in type")
        if val_type.kind == TypeKind.TY_VOID:
            fatal("no QBE void type")

    def qbe_type(self, val_type: DataType) -> str:
        self.check_type(val_type)
        return self.qbe_type_names[val_type.kind.value]

    def qbe_store_type(self, val_type: DataType) -> str:
        self.check_type(val_type)
        return self.qbe_store_type_names[val_type.kind.value]

    def qbe_load_type(self, val_type: DataType) -> str:
        self.check_type(val_type)
        idx = val_type.kind.value
        if val_type.is_unsigned:
            idx += TypeKind.TY_FLT64.value + 1
        return self.qbe_load_type_names[idx]

    def qbe_ext_type(self, val_type: DataType) -> str:
        self.check_type(val_type)
        idx = val_type.kind.value
        if val_type.is_unsigned:
            idx += TypeKind.TY_FLT64.value + 1
        return self.qbe_ext_type_names[idx]

    def gen_temp(self) -> int:
        self.next_temp += 1
        return self.next_temp

    def gen_label(self) -> int:
        self.label_id += 1
        return self.label_id

    def write_all(self, out) -> None:
        out.write(self.output.getvalue())

    def cg_file_preamble(self) -> None:
        pass

    def cg_file_postamble(self) -> None:
        for strlit in strlit_processor:
            self.cg_strlit(strlit)

    def cg_func_preamble(self, fn_sym: Sym) -> None:
        print(f"export function ${fn_sym.name}(", end="", file=self.output)
        for i, param in enumerate(fn_sym.params):
            if i > 0:
                print(", ", end="", file=self.output)
            if param.type.kind == TypeKind.TY_VOID:
                break
            qtype = self.qbe_type(param.type)
            print(f"{qtype} %{param.name}", end="", file=self.output)
        print(") {", file=self.output)
        print("@START", file=self.output)

    def cg_func_postamble(self) -> None:
        print("@END", file=self.output)
        print("  ret", file=self.output)
        print("}", file=self.output)

    def cg_label(self, l: int) -> None:
        print(f"@L{l}", file=self.output)

    def cg_strlit(self, strlit) -> None:
        # 转义特殊字符
        # escaped = strlit.val.replace('\a', '\\a').replace('\b', '\\b')
        # escaped = escaped.replace('\f', '\\f').replace('\n', '\\n')
        # escaped = escaped.replace('\r', '\\r').replace('\t', '\\t').replace('\v', '\\v')
        # escaped = escaped.replace('\\', '\\\\').replace('"', '\\"')
        # print(f"data $L{label} = {{ b \"{escaped}\", b 0 }}", file=self.output)
        label, value = strlit.label, quote_string(strlit.val)
        print(f"data $L{label} = {{ b \"{value}\", b 0 }}", file=self.output)

    def cg_ret(self, node=None):
        if node:
            print(f"  ret {node}", file=self.output)
        else:
            print(f"  ret", file=self.output)

    def cg_jump(self, l: int) -> None:
        print(f"  jmp @L{l}", file=self.output)

    def cg_glob_sym(self, s: Sym) -> None:
        if not s:
            return
        qtype = self.qbe_store_type(s.type)
        if s.type.kind in (TypeKind.TY_FLT32, TypeKind.TY_FLT64):
            prefix, value = qtype + "_", s.init_val.dblval
        else:
            prefix, value = "", s.init_val.intval
        print(f"export data ${s.name} = {{ {qtype} {prefix}{value}, }}", file=self.output)

    def cg_print(self, label: int, temp: int, val_type: DataType) -> None:
        qtype = self.qbe_type(val_type)
        print(f"  call $printf(l $L{label}, {qtype} %.t{temp})", file=self.output)

    def cg_load_lit(self, value: Litval, val_type: DataType) -> int:
        t = self.gen_temp()
        qtype = self.qbe_type(val_type)
        if val_type.kind in (TypeKind.TY_FLT32, TypeKind.TY_FLT64):
            print(f"  %.t{t} ={qtype} copy {qtype}_{value.dblval}", file=self.output)
        else:
            print(f"  %.t{t} ={qtype} copy {value.intval}", file=self.output)
        return t

    def cg_negate(self, t: int, val_type: DataType) -> int:
        t_new = self.gen_temp()
        qtype = self.qbe_type(val_type)
        print(f"  %.t{t_new} ={qtype} neg %.t{t}", file=self.output)
        return t_new

    def cg_not(self, t: int, val_type: DataType) -> int:
        t_new = self.gen_temp()
        print(f"  %.t{t_new} =w not %.t{t}", file=self.output)
        return t_new

    def cg_invert(self, t: int, val_type: DataType) -> int:
        t_new = self.gen_temp()
        qtype = self.qbe_type(val_type)
        print(f"  %.t{t_new} ={qtype} inv %.t{t}", file=self.output)
        return t_new

    def cg_arithmetic(self, op: ASTNodeType, t1: int, t2: int, val_type: DataType) -> int:
        op = self.arithmetic_ops.get(op)
        if not op:
            fatal(f"Unknown arithmetic operator {op}")
        if op == "div" and val_type.is_unsigned:
            op = "divu"
        t_new = self.gen_temp()
        qtype = self.qbe_type(val_type)
        print(f"  %.t{t_new} =w {op} {qtype} %.t{t1}, %.t{t2}", file=self.output)
        return t_new

    def cg_logical(self, op: ASTNodeType, t1: int, t2: int, val_type: DataType) -> int:
        op = self.logical_ops.get(op)
        if not op:
            fatal(f"Unknown logical operator {op}")
        if op == "shr" and val_type.is_unsigned:
            op = "shru"
        t_new = self.gen_temp()
        qtype = self.qbe_type(val_type)
        print(f"  %.t{t_new} =w {op} {qtype} %.t{t1}, %.t{t2}", file=self.output)
        return t_new

    def cg_comparison(self, op: ASTNodeType, t1: int, t2: int, val_type: DataType) -> int:
        op = self.comparison_ops.get(op)
        if not op:
            fatal(f"Unknown comparison operator {op}")
        t_new = self.gen_temp()
        qtype = self.qbe_type(val_type)
        print(f"  %.t{t_new} =w {op} {qtype} %.t{t1}, %.t{t2}", file=self.output)
        return t_new

    def cg_if_false(self, t1: int, label: int) -> None:
        print(f"  jz %.t{t1}, @L{label}", file=self.output)

    def cg_add_local(self, val_type: DataType, sym: Sym) -> None:
        qtype = self.qbe_store_type(val_type)
        print(f"  var {qtype} {sym.name}", file=self.output)

    def cg_load_var(self, sym: Sym) -> int:
        # if sym.type.kind == TypeKind.TY_VOID:
        #     return 0
        t = self.gen_temp()
        qtype = self.qbe_load_type(sym.type)
        print(f"  %.t{t} ={qtype} load ${sym.name}", file=self.output)
        return t

    def cg_stor_var(self, t: int, exprtype: DataType, sym: Sym) -> None:
        qtype = self.qbe_store_type(sym.type)
        print(f"  store {qtype} %.t{t}, ${sym.name}", file=self.output)

    def cg_cast(self, t: int, val_type: DataType, newtype: DataType) -> int:
        t_new = self.gen_temp()
        if val_type.kind in (TypeKind.TY_FLT32, TypeKind.TY_FLT64):
            if newtype.kind in (TypeKind.TY_FLT32, TypeKind.TY_FLT64):
                qtype = self.qbe_type(newtype)
                print(f"  %.t{t_new} ={qtype} copy %.t{t}", file=self.output)
            else:
                qtype = self.qbe_type(val_type)
                new_qtype = self.qbe_type(newtype)
                print(f"  %.t{t_new} ={new_qtype} trunc {qtype} %.t{t}", file=self.output)
        else:
            if newtype.kind in (TypeKind.TY_FLT32, TypeKind.TY_FLT64):
                qtype = self.qbe_type(val_type)
                new_qtype = self.qbe_type(newtype)
                print(f"  %.t{t_new} ={new_qtype} extend {qtype} %.t{t}", file=self.output)
            else:
                qtype = self.qbe_ext_type(val_type)
                new_qtype = self.qbe_type(newtype)
                print(f"  %.t{t_new} ={new_qtype} {qtype} %.t{t}", file=self.output)
        return t_new

    def cg_call(self, sym: Sym, numargs: int, arglist: List[int], typelist: List[DataType]) -> int:
        t = self.gen_temp()
        args = []
        for i in range(numargs):
            qtype = self.qbe_type(typelist[i])
            args.append(f"{qtype} %.t{arglist[i]}")
        args_str = ", ".join(args)
        print(f"  %.t{t} =l call ${sym.name}({args_str})", file=self.output)
        return t


codegen = CodeGenerator()
gen_label = codegen.gen_label
gen_temp = codegen.gen_temp
cg_glob_sym = codegen.cg_glob_sym
cg_negate = codegen.cg_negate
cg_not = codegen.cg_not
cg_invert = codegen.cg_invert
# cg_arithmetic = codegen.cg_arithmetic
# cg_logical = codegen.cg_logical
# cg_comparison = codegen.cg_comparison
# cg_label = codegen.cg_label
# cg_str_lit = codegen.cg_strlit
# cg_ret = codegen.cg_ret
# cg_jump = codegen.cg_jump
# cg_if_false = codegen.cg_if_false
# cg_print = codegen.cg_print
# cg_load_lit = codegen.cg_load_lit
# cg_load_var = codegen.cg_load_var
# cg_stor_var = codegen.cg_stor_var
# cg_cast = codegen.cg_cast
# cg_add_local = codegen.cg_add_local
# cg_call = codegen.cg_call
